tybalt/data_models.py

```
@@ -241,3 +287,205 @@ def combine_models(self, include_labels=False, include_raw=False):
+        module_meanrank_minavg = module_meanrank_minsum / denom
+
+        # We are interested if the features encapsulate gene modules
+        # A lower number across modules indicates a stronger ability to
+        # aggregate genes into features
+        module_min_rank = pd.DataFrame(module_meanrank_df.min(),
+                                       columns=['min_rank'])
+
+        return module_meanrank_df, module_min_rank, module_meanrank_minavg
+
+    def get_group_means(self, df):
+        """
+        Get the mean latent space vector representation of input groups
+        """
+        return df.assign(groups=self.other_df).groupby('groups').mean()
+
+    def get_subtraction(self, group_means, group_list):
+        """
+        Subtract two group means given by group list
+        """
+        a, b = group_list
+
+        a_df = group_means.loc[a, :]
+        b_df = group_means.loc[b, :]
+
+        subtraction = pd.DataFrame(a_df - b_df).T
+
+        return subtraction
+
+    def subtraction_essense(self, group_subtract, mean_rank, node):
+        """
+        Obtain the difference between the subtraction and node of interest
+        """
+        # subset mean rank to node of the "dropped" feature in the simulation
+        feature_essense = mean_rank.loc[:, node]
+
+        # The node essense is the compressed feature with the lowest mean rank
+        node_essense = feature_essense.idxmin()
+        node_idx = int(node_essense.split('_')[1])
+
+        # Ask how different this specific feature is from all others
+        group_z = zscore(group_subtract.iloc[0, :].tolist())
+        node_essense_zscore = group_z[node_idx]
+
+        return node_essense_zscore
+
+    def get_addition(self, group_means, subtraction, group):
+        """
+        Add node to subtraction
+        """
+        mean_feature = group_means.loc[group, :]
+        return subtraction + mean_feature
+
+    def reconstruct_group(self, lsa_result, algorithm=False):
```

Should this throw an error (or warning) if `algorithm` is unsupported?

---

Same comment as before, looks like @danich1 agrees